File: go/parser.go
package main

import (
	"errors"
	"math"
	"strconv"
	"strings"

	"github.com/expr-lang/expr"
)

// errInvalidValor e devolvido pelo parser quando a entrada nao pode ser
// interpretada nem como numero nem como expressao.
var errInvalidValor = errors.New("valor invalido")

// parseValor recebe a string crua vinda do form e retorna o numero
// resultante. Regras:
//   - se comeca com "=", trata o restante como expressao aritmetica
//     (operadores +, -, *, /, parens, decimais com ponto). Avaliado via
//     expr-lang em modo sandboxed.
//   - caso contrario, trata como numero literal aceitando ponto OU virgula
//     como separador decimal.
//
// Retorna erro de validacao com mensagem amigavel em portugues para qualquer
// entrada invalida (sintaxe, divisao por zero, NaN, infinito etc).
func parseValor(raw string) (float64, error) {
	s := strings.TrimSpace(raw)
	if s == "" {
		return 0, errInvalidValor
	}

	if strings.HasPrefix(s, "=") {
		return evalExpression(strings.TrimSpace(s[1:]))
	}
	return parseLiteralNumber(s)
}

// parseLiteralNumber aceita ponto OU virgula como separador decimal.
// Rejeita strings que contenham letras ou multiplos separadores.
func parseLiteralNumber(s string) (float64, error) {
	// normaliza virgula -> ponto. Nao aceita milhar (entrada do usuario nao
	// usa "1.234,56" pra digitar).
	normalized := strings.ReplaceAll(s, ",", ".")
	if strings.Count(normalized, ".") > 1 {
		return 0, errInvalidValor
	}
	v, err := strconv.ParseFloat(normalized, 64)
	if err != nil {
		return 0, errInvalidValor
	}
	if math.IsNaN(v) || math.IsInf(v, 0) {
		return 0, errInvalidValor
	}
	return v, nil
}

// evalExpression compila e executa uma expressao aritmetica usando expr-lang.
// Restringe a sintaxe permitida atraves de validacao previa simples: apenas
// digitos, ponto, espacos e operadores + - * / ( ).
func evalExpression(expression string) (float64, error) {
	if expression == "" {
		return 0, errInvalidValor
	}
	if !allowedExprChars(expression) {
		return 0, errInvalidValor
	}

	program, err := expr.Compile(expression, expr.AsFloat64())
	if err != nil {
		return 0, errInvalidValor
	}
	out, err := expr.Run(program, nil)
	if err != nil {
		// captura divisao por zero, overflow e afins
		return 0, errInvalidValor
	}
	v, ok := out.(float64)
	if !ok {
		return 0, errInvalidValor
	}
	if math.IsNaN(v) || math.IsInf(v, 0) {
		return 0, errInvalidValor
	}
	return v, nil
}

// allowedExprChars garante que a expressao usa apenas o alfabeto permitido:
// digitos, ponto, espaco em branco e operadores aritmeticos basicos.
func allowedExprChars(s string) bool {
	for _, r := range s {
		switch {
		case r >= '0' && r <= '9':
		case r == '.' || r == '+' || r == '-' || r == '*' || r == '/' || r == '(' || r == ')':
		case r == ' ' || r == '\t':
		default:
			return false
		}
	}
	return true
}

File: node/parser.ts
import { Parser } from "expr-eval";

// instancia configurada minimamente (sem funcoes built-in que poderiamos remover,
// mas a lib ja nao expõe `eval` do host — sandbox por design)
const parser = new Parser({
  operators: {
    add: true,
    subtract: true,
    multiply: true,
    divide: true,
    power: false,
    factorial: false,
    remainder: false,
    logical: false,
    comparison: false,
    in: false,
    assignment: false,
    concatenate: false,
  },
});

export interface ParseResult {
  ok: boolean;
  value?: number;
  error?: string;
}

/**
 * Parseia o campo `valor` do form.
 * - Se a string comeca com `=`, avalia como expressao (ponto como decimal).
 * - Caso contrario, aceita numero literal com ponto OU virgula como decimal.
 * Retorna {ok:true, value} ou {ok:false, error}.
 */
export function parseValor(raw: string): ParseResult {
  const trimmed = (raw ?? "").trim();
  if (trimmed === "") {
    return { ok: false, error: "informe um valor" };
  }

  if (trimmed.startsWith("=")) {
    const expr = trimmed.slice(1).trim();
    if (expr === "") {
      return { ok: false, error: "expressao vazia apos =" };
    }
    // por convencao da spec, virgula nao e aceita em expressao
    if (expr.includes(",")) {
      return { ok: false, error: "use ponto como decimal em expressoes" };
    }
    try {
      const compiled = parser.parse(expr);
      const value = compiled.evaluate({}); // contexto vazio = sem variaveis
      if (typeof value !== "number" || !Number.isFinite(value)) {
        return { ok: false, error: "expressao nao gerou numero finito" };
      }
      return { ok: true, value };
    } catch (err) {
      const msg = err instanceof Error ? err.message : "erro de sintaxe";
      return { ok: false, error: `expressao invalida: ${msg}` };
    }
  }

  // numero literal: aceita ponto OU virgula como decimal
  // remove separadores de milhar tipo "1.234,56" -> "1234.56"
  // estrategia: se tem virgula, virgula e decimal e ponto e milhar; senao ponto e decimal.
  let normalized: string;
  if (trimmed.includes(",")) {
    normalized = trimmed.replace(/\./g, "").replace(",", ".");
  } else {
    normalized = trimmed;
  }
  const n = Number(normalized);
  if (!Number.isFinite(n)) {
    return { ok: false, error: "valor invalido" };
  }
  return { ok: true, value: n };
}
